Add a page documenting Godot's internal rendering architecture (#6133)
godotengine/godot-docs: contributing/development/core_and_modules/internal_rendering_architecture.rst

.. _doc_internal_rendering_architecture:

Internal rendering architecture
===============================

This page is a high-level overview of Godot 4's internal renderer design.
It does not apply to previous Godot versions.

The goal of this page is to document design decisions taken to best suit
:ref:`Godot's design philosophy <doc_best_practices_for_engine_contributors>`,
while providing a starting point for new rendering contributors.

If you have questions about rendering internals not answered here, feel free to
ask in the ``#rendering`` channel of the
`Godot Contributors Chat <https://chat.godotengine.org/channel/rendering>`__.

.. note::

    If you have difficulty understanding concepts on this page, it is
    recommended to go through an OpenGL tutorial such as
    `LearnOpenGL <https://learnopengl.com/>`__.

    Modern low-level APIs (Vulkan/Direct3D 12) require intermediate
    knowledge of higher-level APIs (OpenGL/Direct3D 11) to be used
    effectively. Many concepts are common to both kinds of graphics APIs.
    Once you are familiar with modern OpenGL, you may follow a Vulkan tutorial
    such as `Vulkan Tutorial <https://vulkan-tutorial.com/>`__ or

>     `Vulkan Guide <https://vkguide.dev/>`__.

Author: On the other hand, do we expect people to know about concepts used in modern low-level APIs (barriers, push constants, specialization constants, etc)? RenderingDevice requires knowledge of those to use the API effectively.

>     `Vulkan Guide <https://vkguide.dev/>`__.

Reviewer: You need to understand many of those concepts to make best use of most APIs (modern OpenGL, WebGPU, Metal, D3D10, D3D11).

The point is, we shouldn't build an expectation that users should know how to use Vulkan in order to contribute to Godot's renderer. RenderingDevice is way easier to use than Vulkan and it requires no Vulkan-specific knowledge. We already have tons of users saying they want to contribute, but don't know Vulkan so they don't.

Rendering drivers
-----------------

> Godot 4 supports the following graphics APIs:

Author: I've put the rendering drivers section after the rendering methods one.

done by running the vertex shader corresponding to the previous rendered frame
(with the previous camera transform) in addition to the vertex shader for the
current rendered frame, then storing the difference between them in a color buffer.

Using `FSR 2.0 <https://github.com/GPUOpen-Effects/FidelityFX-FSR2>`__ instead
of a custom TAA implementation is planned in a future release.

**TAA resolve:**

- `servers/rendering/renderer_rd/shaders/effects/taa_resolve.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/effects/taa_resolve.glsl>`__

Global illumination
^^^^^^^^^^^^^^^^^^^

.. note::

  VoxelGI and SDFGI are only available in the Forward+ backend, not the
  Forward Mobile or Compatibility methods.

  LightmapGI *baking* is only available in the Forward+ and Forward
  Mobile methods, and can only be performed within the editor (not in an
  exported project). LightmapGI *rendering* is available in all methods
  including Compatibility.

Godot supports voxel-based GI (VoxelGI), signed distance field GI (SDFGI) and
lightmap baking and rendering (LightmapGI). These techniques can be used
simultaneously if desired.

> Lightmap baking happens on the GPU using Vulkan compute shaders – CUDA or OpenCL
> are **not** used. Using Vulkan compute allows for greater compatibility and
> performance on all kinds of GPUs. In contrast, CUDA only runs on NVIDIA hardware
> and OpenCL performance is crippled on NVIDIA GPUs.
> 
> The GPU-based lightmapper is implemented in the LightmapperRD class, which inherits
> from the Lightmapper class. This allows for implementing additional lightmappers,
> paving the way for a future port of the CPU-based lightmapper present in Godot 3.x.
> This would allow baking lightmaps while using the Compatibility backend.

Reviewer: ```suggestion
Lightmap baking happens on the GPU using Vulkan compute shaders. The
GPU-based lightmapper is implemented in the LightmapperRD class, which inherits
from the Lightmapper class. This allows for implementing additional lightmappers,
paving the way for a future port of the CPU-based lightmapper present in Godot 3.x.
This would allow baking lightmaps while using the Compatibility backend.
```

> This would allow baking lightmaps while using the Compatibility backend.

Author: It could be worth adding a brief paragraph on why Vulkan compute is used instead of CUDA/OpenCL (which might seem like the logical choices for doing this at first), but I think shortening this makes sense.

> This would allow baking lightmaps while using the Compatibility backend.

Reviewer: Well, the main reason is just that we already have a vulkan context and it is way faster to use Vulkan compute then to take Vulkan GPU resources, upload them to CPU, download them back to GPU for CUDA, process on CUDA, upload back to CPU, put them in vulkan compatible format, download back to GPU for Vulkan.

That being said, in my opinion, a discussion on CUDA vs. Vulkan is not suitable for this document. CUDA isn't a viable choice for games. Our choice to not use it is not unique and does not require an explanation

**Depth of field GLSL shader (compute - used for Forward+):**

- `servers/rendering/renderer_rd/shaders/effects/bokeh_dof.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/effects/bokeh_dof.glsl>`__

**Depth of field GLSL shader (raster - used for Forward Mobile):**

- `servers/rendering/renderer_rd/shaders/effects/bokeh_dof_raster.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/effects/bokeh_dof_raster.glsl>`__

Screen-space effects (SSAO, SSIL, SSR, SSS)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note::

  Only available in the Forward+ backend, not the Forward Mobile or
  Compatibility methods.

The Forward+ backend supports screen-space ambient occlusion,
screen-space indirect lighting, screen-space reflections and subsurface scattering.

SSAO uses an implementation derived from Intel's
`ASSAO <https://www.intel.com/content/www/us/en/developer/articles/technical/adaptive-screen-space-ambient-occlusion.html>`__
(converted to Vulkan). SSIL is derived from SSAO to provide high-performance
indirect lighting.

> When both SSAO and SSIL are enabled, parts of SSAO and SSIL are run in parallel
> to reduce the performance impact.

Reviewer: ```suggestion
When both SSAO and SSIL are enabled, parts of SSAO and SSIL are shared
to reduce the performance impact.
```

While modern GPUs can handle drawing a lot of triangles, the number of draw
calls in complex scenes can still be a bottleneck (even with Vulkan and Direct3D
12).

Godot 4 supports occlusion culling to reduce overdraw (when the depth prepass
is disabled) and reduce vertex throughput.
This is done by rasterizing a low-resolution buffer on the CPU using
`Embree <https://github.com/embree/embree>`__. The buffer's resolution depends
on the number of CPU threads on the system, as this is done in parallel.
This buffer includes occluder shapes that were baked in the editor or created at
run-time.

As complex occluders can introduce a lot of strain on the CPU, baked occluders
can be simplified automatically when generated in the editor.

Godot's occlusion culling doesn't support dynamic occluders yet, but
OccluderInstance3D nodes can still have their visibility toggled or be moved.
However, this will be slow when updating complex occluders this way. Therefore,
updating occluders at run-time is best done only on simple occluder shapes such
as quads or cuboids.

This CPU-based approach has a few advantages over other solutions, such as
portals and rooms or a GPU-based culling solution:

> - No manual setup required.

Reviewer: ```suggestion
- No manual setup required (but can be tweaked manually for best performance).
```
